sensor: Add support for INA219

leonardp opened the pull request:
This adds support and an example application for the TI INA219 zero-drift, bidirectional current/power monitor.

There already is a "quick and dirty" application example using this device [here](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/current_sensing).

Improvements over the old example application are:
- usage of the sensor API
- configuration/calibration data via devicetree
- reverse current now working
- power management
- triggered mode to save on power

Signed-off-by: Leonard Pollak leonardp@tr-host.de

mbolivar-nordic commented on dts/bindings/sensor/ti,ina219.yaml:
```diff
@@ -0,0 +1,107 @@
+    lsb-microamp:
+      type: int
+      required: false
+      default: 100
```
Justifications for all defaults must be provided; please see https://docs.zephyrproject.org/latest/guides/dts/bindings.html#default-values-for-properties for the rules.

leonardp replied:
Good catch! It doesn't seem right to have a default value for this. I will change it accordingly.

mbolivar-nordic replied:
OK, thanks! Please note that the requirements in that link apply to all defaults in this PR

leonardp replied:
I pushed the change. All other defaults correspond to the default power-up state of the config register.

mbolivar-nordic replied:
> All other defaults correspond to the default power-up state of the config register.

Thanks for confirming that, but you need to explicitly document that in the binding. From the bindings link I posted (emphasis mine):

"In any case where default: is used, the property documentation should **explain why the value was selected** and any conditions that would make it necessary to provide a different value. (**This is mandatory for built-in bindings.**)"

leonardp replied:
Sorry for the inconvenience but I didn't read up to that paragraph.
I have now added (what I think is) all the necessary documentation.

mbolivar-nordic reviewed (changes requested):
Commenting on the binding only.

Commit:
samples: sensor: added sample app for INA219

This adds an example application for the TI INA219 Zero-Drift,
Bidirectional Current/Power Monitor with I2C Interface

Signed-off-by: Leonard Pollak <leonardp@tr-host.de>

Commit:
drivers: sensor: INA219: added support

This adds support for the TI INA219 Zero-Drift, Bidirectional
Current/Power Monitor with I2C Interface

Signed-off-by: Leonard Pollak <leonardp@tr-host.de>

mbolivar-nordic reviewed (approved):
Thanks. The binding looks good to me now. I didn't review the rest in detail.

mbolivar-nordic reviewed (commented):
While working on something else, I discovered samples/drivers/current_sensing, which is a pretty hacky example of getting this exact sensor working on a particular breakout board with a hardcoded I2C device name.

I think it's no longer necessary with this patch -- @leonardp could you please add a commit to this pull request removing the stale sample?

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,353 @@
+			&reg_addr, sizeof(reg_addr),
+			rx_buf, sizeof(rx_buf));
+
+	*reg_data = ((uint16_t)rx_buf[0] << 8 | rx_buf[1]);
```
`byteorder.h` API should be used

leonardp replied:
I have done this for both the read and write functions.

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,353 @@
+	const struct ina219_config *cfg = dev->config;
+	uint16_t cal;
+
+	cal = 40960000U / ((cfg->r_shunt) * (cfg->current_lsb));
```
constant should be defined

leonardp replied:
this value is now defined as a constant

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,353 @@
+				    INA219_REG_CONF,
+				    INA219_MODE_MASK,
+				    INA219_MODE_NORMAL) ) {
+		LOG_DBG("%s: Failed to start measurement.", dev->name);
```
`LOG_ERR`

leonardp replied:
changed it to LOG_ERR

galak reviewed (dismissed):
please add to tests/drivers/build_all/sensor/

Commit:
tests: drivers: added INA219 to build tests

This adds the INA219 driver to the build tests.

Signed-off-by: Leonard Pollak <leonardp@tr-host.de>

leonardp commented:
> please add to tests/drivers/build_all/sensor/

Is this sufficient to do that? https://github.com/zephyrproject-rtos/zephyr/pull/35538/commits/329ace230ad629fca5c34f1de14861cee8af1ef9


@mbolivar-nordic Thanks for reviewing the bindings. I have added https://github.com/zephyrproject-rtos/zephyr/pull/35538/commits/0472698acbf06522f1468be5726405d1d6c86849 to remove the old sample code.

mbolivar-nordic commented:
>  I have added [0472698](https://github.com/zephyrproject-rtos/zephyr/commit/0472698acbf06522f1468be5726405d1d6c86849) to remove the old sample code.

Thanks!

jfischer-no commented on samples/sensor/ina219/src/main.c:
```diff
@@ -0,0 +1,54 @@
+
+void main(void)
+{
+#if defined(CONFIG_USB_UART_CONSOLE)
```
Please remove USB device support, it is not necessary here.

leonardp replied:
There is an overlay-usb.conf that can be used to enable USB device support.
Should I remove that one as well?

jfischer-no replied:
yes, please.

leonardp replied:
I have removed all USB related stuff from the sample code.

leonardp commented on drivers/sensor/ina219/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2021 Leonard Pollak
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig INA219
```
i've updated the Kconfig accordingly

gmarull commented on samples/sensor/ina219/src/main.c:
```diff
@@ -0,0 +1,43 @@
+
+void main(void)
+{
+	const struct device *ina = device_get_binding("INA219");
+	struct sensor_value v_shunt, v_bus, power, current;
+
+	if (ina == NULL) {
+		printf("Could not get INA219 device\n");
+		return;
+	}
```
better use `DEVICE_DT_GET_ANY(ti_ina219)` (check other samples)

leonardp replied:
updated

gmarull commented on samples/sensor/ina219/src/main.c:
```diff
@@ -0,0 +1,43 @@
+		sensor_channel_get(ina, SENSOR_CHAN_BUS_POWER, &power);
+		sensor_channel_get(ina, SENSOR_CHAN_BUS_CURRENT, &current);
+
+
```
nit: extra blank line

leonardp replied:
black line removed

gmarull commented on samples/sensor/ina219/src/main.c:
```diff
@@ -0,0 +1,43 @@
+	while (true) {
+
+		if (sensor_sample_fetch(ina)){
+			printf("Could not fetch sensor data.\n");
```
code will continue with `get` even if `fetch` fails

leonardp replied:
added a return

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+{
+	uint16_t reg_data;
+
+	if (ina219_reg_read(dev, addr, &reg_data)) {
+		return -EIO;
+	}
```
error code should be propagated here

leonardp replied:
updated

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+	uint16_t status;
+	uint16_t tmp;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
```
I'd return -ENOTSUP if given channel is not supported.

leonardp replied:
i have updated the ina219_sample_fetch function to also be able to only fetch a single (private) channel.
not sure if it good the way i did it...

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+#if CONFIG_PM_DEVICE
+	if (!(data->pm_state == PM_DEVICE_STATE_ACTIVE)) {
```
`!=`

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+#endif
+
+	/* Trigger measurement and wait for completion */
+	if (ina219_reg_field_update(dev,
+				    INA219_REG_CONF,
+				    INA219_MODE_MASK,
+				    INA219_MODE_NORMAL)) {
+		LOG_ERR("%s: Failed to start measurement.", dev->name);
+		return -EIO;
+	}
```
error code should be propagated

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+
+	/* Check for overflow */
+	if (INA219_OVF_STATUS(status)) {
+		LOG_DBG("%s: Power and/or Current calculations are out of range.", dev->name);
```
question: is this a situation that should report error or just log?

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+	double tmp;
+	int8_t sign = 1;
+
+	switch ((int16_t)chan) {
```
why casting to `int16_t`?

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+
+#ifdef CONFIG_PM_DEVICE
+static int ina219_set_power_state(const struct device *dev,
+				  uint32_t power_state)
```
`enum pm_device_state` must be used now

gmarull replied:
not fixed

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+	}
+
+	if (power_state == PM_DEVICE_STATE_ACTIVE) {
+		reg_val = INA219_MODE_NORMAL;
```
this assignment is redundant here, since the function returns afterwards.

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+		return 0;
+	}
+
+	if (ina219_reg_field_update(dev,
+				INA219_REG_CONF,
+				INA219_MODE_MASK,
+				reg_val)) {
+		LOG_DBG("%s: Failed to set power state.", dev->name);
+		return -EIO;
+	}
```
error code should be propagated

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+
+static int ina219_pm_ctrl(const struct device *dev,
+	uint32_t ctrl_command,
+	uint32_t *state,
```
`enum pm_device_state`

leonardp replied:
updated

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,358 @@
+		return -ENODEV;
+	}
+
+	if (ina219_reg_write(dev, INA219_REG_CONF, INA219_RST)){
+		LOG_DBG("%s: Could not reset device.", dev->name);
+		return -ENODEV;
+	}
+
+	if (ina219_set_config(dev)) {
+		LOG_DBG("%s: Could not set configuration data.", dev->name);
+		return -ENODEV;
+	}
+
+	if (ina219_set_calib(dev)) {
+		LOG_DBG("%s: Could not set calibration data.", dev->name);
+		return -ENODEV;
+	}
+
+	/* Set measurement delay */
+	if (ina219_set_msr_delay(dev)) {
+		LOG_ERR("%s: Could not get measurement delay.", dev->name);
+		return -EINVAL;
+	}
```
error codes should be propagated

leonardp replied:
updated

gmarull commented on drivers/sensor/ina219/ina219.h:
```diff
@@ -0,0 +1,108 @@
+
+/* Others */
+#define INA219_SIGN_BIT(x)	((x) >> 15) & 0x1
+#define INA219_V_BUS_MUL	0.004
+#define INA219_SI_MUL		0.00001
```
nit: `f` suffix

leonardp replied:
updated

gmarull commented on drivers/sensor/ina219/ina219.h:
```diff
@@ -0,0 +1,108 @@
+	uint16_t calib;
+	uint32_t msr_delay;
+#ifdef CONFIG_PM_DEVICE
+	uint32_t pm_state;
```
`enum pm_device_state`

leonardp replied:
updated

gmarull commented on include/drivers/sensor/ina219.h:
```diff
@@ -0,0 +1,36 @@
+
+enum sensor_channel_ina219 {
+	/** Shunt Voltage in V **/
+	SENSOR_CHAN_V_SHUNT = SENSOR_CHAN_PRIV_START,
+	/** Bus Voltage in V **/
+	SENSOR_CHAN_BUS_VOLTAGE,
+	/** Power **/
+	SENSOR_CHAN_BUS_POWER,
+	/** Current **/
+	SENSOR_CHAN_BUS_CURRENT,
```
would probably be a good idea to namespace these channels, ie, `SENSOR_CHAN_INA219_BUS_VOLTAGE`

leonardp replied:
i prefixed the channels as you suggested and updated the relevant parts of the code

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,388 @@
+		return -ENOTSUP;
+	}
+
+#if CONFIG_PM_DEVICE
+	if (data->pm_state != PM_DEVICE_STATE_ACTIVE) {
+		LOG_DBG("Device inactive.");
+		return -EBUSY;
+	}
+#endif
```
devices that need PM at runtime should implement the runtime PM API, so this is not required. There's currently ongoing work to improve the device PM state on Zephyr, ref #36125

leonardp replied:
removed this section. nice to see work being done to improve the PM API, really looking forward to this.

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,388 @@
+		LOG_DBG("Power and/or Current calculations are out of range.");
+	}
+
+	if (chan != SENSOR_CHAN_ALL ||
+		chan_p != SENSOR_CHAN_INA219_V_SHUNT) {
```
I'm a bit confused here, but shouldn't that be:

```
	if (chan == SENSOR_CHAN_ALL ||
		chan_p == SENSOR_CHAN_INA219_V_SHUNT) {
```

leonardp replied:
yes, i am also confused. changed it to your suggestion :+1:

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,388 @@
+	}
+
+	/* Check for overflow */
+	if (INA219_OVF_STATUS(status)) {
+		LOG_DBG("Power and/or Current calculations are out of range.");
+	}
```
This should return an error, obtained data may be meaningless:

![image](https://user-images.githubusercontent.com/25011557/125253925-3ee79d80-e2fa-11eb-8e23-fec512cf57e4.png)

leonardp replied:
i've changed it to `LOG_ERR`, although i am not 100% sure about it being an actual error.
it's a case of: "3.6 röntgen -- not great, not terrible"...

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,381 @@
+
+	/* Check for overflow */
+	if (INA219_OVF_STATUS(status)) {
+		LOG_ERR("Power and/or Current calculations are out of range.");
```
If it's an acceptable case, then I'd change to `LOG_WRN`

leonardp replied:
i did just that. i think it is fine this way.
thanks for all the reviewing

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,381 @@
+static int ina219_pm_ctrl(const struct device *dev,
+	uint32_t ctrl_command,
+	enum pm_device_state *state,
+	pm_device_cb cb,
+	void *arg)
+{
+	int rc = 0;
+
+	if (ctrl_command == PM_DEVICE_STATE_SET) {
+		rc = ina219_set_power_state(dev, *state);
+	} else if (ctrl_command == PM_DEVICE_STATE_GET) {
+		rc = ina219_get_power_state(dev, state);
+	}
+
+	if (cb) {
+		cb(dev, rc, state, arg);
+	}
```
callback has been removed, see https://github.com/zephyrproject-rtos/zephyr/pull/36757

leonardp replied:
i like it.
I have removed the `pm_device_cb` argument as well as the if block

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,377 @@
+
+static int ina219_pm_ctrl(const struct device *dev,
+	uint32_t ctrl_command,
+	enum pm_device_state *state,
+	void *arg)
```
```suggestion
	enum pm_device_state *state)
```

leonardp replied:
updated

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,377 @@
+		return 0;
+	}
+
+	if (ina219_reg_field_update(dev,
+				INA219_REG_CONF,
+				INA219_MODE_MASK,
+				reg_val)) {
+		LOG_DBG("Failed to set power state.");
+		return -EIO;
+	}
```
error code needs to be propagated

leonardp replied:
updated

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,377 @@
+	} else if (power_state == PM_DEVICE_STATE_OFF) {
+		reg_val = INA219_MODE_OFF;
+	} else {
+		return 0;
```
```suggestion
		return -ENOTSUP;
```

leonardp replied:
updated

leonardp commented:
The documentation build times out with:
```
Warning, treated as error:
/home/runner/work/zephyr/zephyr/doc/_build/src/samples/sensor/ina219/README.rst:16:'any' reference target not found: https://www.ti.com/product/INA219
Error: The operation was canceled.
```
however the link is working fine for me

gmarull commented on samples/sensor/ina219/README.rst:
```diff
@@ -0,0 +1,46 @@
+References
+**********
+
+ - `INA219 sensor <https://www.ti.com/product/INA219>`
```
This is missing a trailing underscore, 
```rst
`text <url>`_
```

leonardp replied:
thanks I've updated it. RST has a bit of a weird syntax imo.
now the tests should run fine.

leonardp commented:
@galak can you please revisit?
i would like to see this merged

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,378 @@
+	return 0;
+}
+
+static int ina219_pm_ctrl(const struct device *dev,
```
PM code should be adjusted following latest changes, see https://github.com/zephyrproject-rtos/zephyr/pull/36759

leonardp replied:
updated accordingly

leonardp commented:
I have removed the private channels and replaced them with the public ones introduced by https://github.com/zephyrproject-rtos/zephyr/pull/35753
The channel for the shunt voltage was therefore removed completely.

gmarull commented on samples/sensor/ina219/src/main.c:
```diff
@@ -0,0 +1,47 @@
+#include <drivers/sensor.h>
+
+
+#if !DT_HAS_COMPAT_STATUS_OKAY(ti_ina219)
+#error "No ti,ina219 compatible node found in the device tree"
+#endif
```
better use `DEVICE_DT_GET_ONE` so that you get a compilation error if there's no device.

leonardp replied:
updated.

gmarull commented on samples/sensor/ina219/src/main.c:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	while (true) {
+
```
nit: extra blank line

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,315 @@
+		(cfg->sadc & INA219_ADC_MASK) << INA219_SADC_SHIFT |
+		(cfg->mode & INA219_MODE_NORMAL);
+
+
```
nit: extra blank line

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,315 @@
+#ifdef CONFIG_PM_DEVICE
+static int ina219_pm_ctrl(const struct device *dev, enum pm_device_action action)
+{
+	int rc = 0;
```
not used

leonardp replied:
removed

gmarull commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -0,0 +1,315 @@
+	static struct ina219_data ina219_data_##n;		\
+								\
+	static const struct ina219_config ina219_config_##n = {	\
+		.bus = DEVICE_DT_GET(DT_INST_BUS(n)),		\
+		.i2c_addr = DT_INST_REG_ADDR(n),		\
```
use `i2c_dt_spec`: https://github.com/zephyrproject-rtos/zephyr/pull/37359

leonardp replied:
updated the driver accordingly.
I could also update the sgp40 and sht4x driver and add it to this pull request.
Would that be OK?

gmarull replied:
better open another PR for that, would be a nice improvement :-)

leonardp replied:
see: https://github.com/zephyrproject-rtos/zephyr/pull/37741

gmarull reviewed (changes requested):
Some more changes requested, please check my comments. I still see that samples comes before the actual driver, can you re-order commits in proper chronological order?

MaureenHelm reviewed (changes requested):
Please rebase and resolve the merge conflict

Commit:
samples: drivers: removed old sample code for INA219

This removes the now stale sample code for the INA219

Signed-off-by: Leonard Pollak <leonardp@tr-host.de>

leonardp commented on samples/sensor/ina219/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
rebased and updated
